unit_tests/test_ops_openstack.py

```diff
@@ -0,0 +1,268 @@
+#!/usr/bin/env python3
+
+# Copyright 2020 Canonical Ltd.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+# http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import unittest
+
+from mock import patch, MagicMock
+
+from ops.testing import Harness
+from ops.model import (
+    ActiveStatus,
+    BlockedStatus,
+    MaintenanceStatus,
+    WaitingStatus,
+)
+
+import ops_openstack
+
+
+class OpenStackTestAPICharm(ops_openstack.OSBaseCharm):
+
+    PACKAGES = ['keystone-common']
+    REQUIRED_RELATIONS = ['shared-db']
+    RESTART_MAP = {
+        '/etc/f1.conf': ['apache2'],
+        '/etc/f2.conf': ['apache2', 'ks-api'],
+        '/etc/f3.conf': []}
+
+    def custom_status_check(self):
+        if self.model.config.get('custom-check-fail', 'False') == 'True':
+            self.unit.status = MaintenanceStatus(
+                'Custom check failed')
+            return False
+        else:
+            return True
+
+
+class CharmTestCase(unittest.TestCase):
+
+    def setUp(self, obj, patches):
+        super(CharmTestCase, self).setUp()
+        self.patches = patches
+        self.obj = obj
+        self.patch_all()
+
+    def patch(self, method):
+        _m = patch.object(self.obj, method)
+        mock = _m.start()
+        self.addCleanup(_m.stop)
+        return mock
+
+    def patch_all(self):
+        for method in self.patches:
+            setattr(self, method, self.patch(method))
+
+
+class TestOSBaseCharm(CharmTestCase):
+
+    PATCHES = [
+        'add_source',
+        'apt_update',
+        'apt_install',
+        'os_utils']
+
+    def setUp(self):
+        super(TestOSBaseCharm, self).setUp(ops_openstack, self.PATCHES)
+        self.os_utils.manage_payload_services = MagicMock()
+        self.harness = Harness(
+            OpenStackTestAPICharm,
+            meta='''
+                name: client
+                provides:
+                  ceph-client:
+                    interface: ceph-client
+            ''',
+            actions='''
+                pause:
+                    description: pause action
+                resume:
+                    description: resume action
+            ''')
+
+    def test_init(self):
+        self.harness.begin()
+        self.assertFalse(self.harness.charm.state.is_started)
+        self.assertFalse(self.harness.charm.state.is_paused)
+        self.assertFalse(self.harness.charm.state.series_upgrade)
+        events = list(self.harness.charm.on.events())
+        self.assertEqual(
+            sorted(events),
```

Comment:
This feels a little excessive because the framework tests are responsible for checking that events are dynamically defined.